src/main/java/net/fabricmc/loader/discovery/ModResolver.java

```diff
@@ -314,46 +314,117 @@ private static IVecInt toVecInt(IntStream stream) {
+				.append(dependency.getModId()).append(".");
+	}
+
+	private void appendUnsatisfiedDependencyError(StringBuilder errors, ModDependency dependency, ModCandidate depCandidate) {
+		errors.append("but a non-matching version is present: ").append(getCandidateFriendlyVersion(depCandidate)).append("!");
+		errors.append("\n\t - You must install ").append(getDependencyVersionRequirements(dependency)).append(" of ")
+				.append(getCandidateName(depCandidate)).append(".");
+	}
+
+	private void appendConflictError(StringBuilder errors, ModCandidate candidate, ModCandidate depCandidate) {
+		final String depCandidateVer = getCandidateFriendlyVersion(depCandidate);
+		errors.append("but a matching version is present: ").append(depCandidateVer).append("!");
+		errors.append("\n\t - While this won't prevent you from starting the game,");
+		errors.append(" the developer(s) of ").append(getCandidateName(candidate));
+		errors.append(" have found that version ").append(depCandidateVer).append(" of ").append(getCandidateName(depCandidate));
+		errors.append(" conflicts with their mod.");
+		errors.append("\n\t - It is heavily recommended to remove one of the mods.");
+	}
+
+	private void appendBreakingError(StringBuilder errors, ModCandidate candidate, ModCandidate depCandidate) {
+		final String depCandidateVer = getCandidateFriendlyVersion(depCandidate);
+		errors.append("but a matching version is present: ").append(depCandidate.getInfo().getVersion()).append("!");
+		errors.append("\n\t - The developer(s) of ").append(getCandidateName(candidate));
+		errors.append(" have found that version ").append(depCandidateVer).append(" of ").append(getCandidateName(depCandidate));
+		errors.append(" critically conflicts with their mod.");
+		errors.append("\n\t - You must remove one of the mods.");
+	}
+
+	private static String getCandidateName(ModCandidate candidate) {
+		return "'" + candidate.getInfo().getName() + "' (" + candidate.getInfo().getId() + ")";
+	}
+
+	private static String getCandidateFriendlyVersion(ModCandidate candidate) {
+		return candidate.getInfo().getVersion().getFriendlyString();
+	}
+
+	private static String getDependencyVersionRequirements(ModDependency dependency) {
+		return dependency.getVersionRequirements().stream().map(predicate -> {
+			String version = predicate.getVersion();
+			String[] parts;
+			switch(predicate.getType()) {
+			case ANY:
+				return "any version";
+			case EQUALS:
+				return "version " + version;
+			case GREATER_THAN:
+				return "any version after " + version;
+			case LESSER_THAN:
+				return "any version before " + version;
+			case GREATER_THAN_OR_EQUAL:
+				return "version " + version + " or later";
+			case LESSER_THAN_OR_EQUAL:
+				return "version " + version + " or earlier";
+			case SAME_MAJOR:
+				parts = version.split("\\.");
+
+				for (int i = 1; i < parts.length; i++) {
+					parts[i] = "x";
+				}
+
+				return "version " + String.join(".", parts);
```

Comment:
The caret (`^`) ranges actually check that the left-most non-zero digit matches, so this would have different behaviour for versions such as 0.a.b and 0.0.c. (This is detailed in the [NPM semver docs](https://docs.npmjs.com/misc/semver#caret-ranges-123-025-004), which are linked in the fabric.mod.json spec as the definition of the dep strings.)

Reply:
Oops. `~` is probably wrong too, then...

Reply:
What should the equivalent values in the `Type` enum be called, then? Just `CARET` and `TILDE`?

Reply:
According to Player, this special condition doesn't exist here. Marking as resolved.